### PR TITLE
Tpetra:  added methods to replace Domain and Range maps in CrsMatrix and CrsGraph

### PR DESCRIPTION
**Pull request by kddevin**
@trilinos/tpetra 

## Motivation

Implementing replaceDomainMap, replaceRangeMap, replaceRangeMapAndExporter to complement existing replaceDomainMapAndImporter, as recommended in #9451 

Unfortunately, these methods do not resolve the issue #9391 (creating a new matrix with the same local graph, rowMap and columnMap as an existing matrix, but different Range and Domain maps).  See further notes in re-opened issue #9391.

## Testing

New tests added to CrsMatrix_replaceDomainMapAndImporter.cpp to exercise new replaceDomainMap (both with a different map and the same map).
New file of tests CrsMatrix_replaceRangeMapAndExporter.cpp to exercise new replaceRangeMap (both with a different map and the same map) and replaceRangeMapAndExporter.
Tested on mac with gcc 8.5 and openmpi 1.10.7.

### CODE REVIEW

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp [hunk 1]
```diff
@@ -98,16 +98,49 @@ namespace {
...
+    (B).apply(BVecX,BVecY); \
+    \
+    BVecY.update (-STS::one (), AVecY, STS::one ()); \
+    norm = BVecY.norm2(); \
+    \
+    out << "Residual 2-norm: " << norm << endl \
+        << "Residual 1-norm: " << BVecY.norm1 () << endl \
+        << "Residual Inf-norm: " << BVecY.normInf () << endl; \
+    \
+    /* Macros don't like spaces, so we put the test outside the */ \
```
Comment (tasmith4) [inline packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp:128]: are these comments still relevant now that the code has been rearranged?

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp [hunk 2]
```diff
@@ -146,59 +181,135 @@ namespace {
...
+    const global_size_t INVALID = OrdinalTraits<global_size_t>::invalid();
+    // get a comm
+    RCP<const Comm<int> > comm = Tpetra::getDefaultComm();
+    const size_t numImages = comm->getSize();
+    const size_t myImageID = comm->getRank();
+    if (numImages < 3) return;
+    // create a Map
+    RCP<const Map<LO,GO,Node> > map = createContigMapWithNode<LO,GO,Node>(INVALID,ONE,comm);
+
+    // RCP<FancyOStream> fos = Teuchos::fancyOStream(rcp(&std::cout,false));
```
Comment (tasmith4) [inline packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp:215]: seems like a leftover line that can be deleted?

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp [hunk 3]
```diff
@@ -146,59 +181,135 @@ namespace {
...
+    // get a comm
+    RCP<const Comm<int> > comm = Tpetra::getDefaultComm();
+    const size_t numImages = comm->getSize();
+    const size_t myImageID = comm->getRank();
+    if (numImages < 3) return;
+    // create a Map
+    RCP<const Map<LO,GO,Node> > map = 
+              createContigMapWithNode<LO,GO,Node>(INVALID,2,comm);
+
+    // RCP<FancyOStream> fos = Teuchos::fancyOStream(rcp(&std::cout,false));
```
Comment (tasmith4) [inline packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp:271]: same comment as above

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+  TEUCHOS_STATIC_SETUP()
+  {
+    Teuchos::CommandLineProcessor &clp = Teuchos::UnitTestRepository::getCLP();
+    clp.addOutputSetupOptions(true);
+  }
+
+  //
+  // UNIT TESTS
+  //
+#define applyAndCheckResult( A, B, newMap) \
```
Comment (tasmith4) [inline packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp:101]: same comment on this function as in the previous file.  Also, if it's repeated across separate files, shouldn't it go in a header to reduce duplication?

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    const global_size_t INVALID = OrdinalTraits<global_size_t>::invalid();
+    // get a comm
+    RCP<const Comm<int> > comm = Tpetra::getDefaultComm();
+    const size_t numImages = comm->getSize();
+    const size_t myImageID = comm->getRank();
+    if (numImages < 3) return;
+    // create a Map
+    RCP<const Map<LO,GO,Node> > map = createContigMapWithNode<LO,GO,Node>(INVALID,ONE,comm);
+
+    // RCP<FancyOStream> fos = Teuchos::fancyOStream(rcp(&std::cout,false));
```
Comment (tasmith4) [inline packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp:156]: same comment as in previous file

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
```
Comment (tasmith4) [inline packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp:215]: same comment as above

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+    // get a comm
+    RCP<const Comm<int> > comm = Tpetra::getDefaultComm();
+    const size_t numImages = comm->getSize();
+    const size_t myImageID = comm->getRank();
+    if (numImages < 3) return;
+    // create a Map
+    RCP<const Map<LO,GO,Node> > map = 
+              createContigMapWithNode<LO,GO,Node>(INVALID,2,comm);
+
+    // RCP<FancyOStream> fos = Teuchos::fancyOStream(rcp(&std::cout,false));
```
Comment (tasmith4) [inline packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp:271]: same comment as above

**Comment by tasmith4** [comment]
@kddevin, the PR description mentions additional notes in re-opened issue #9391 -- somehow I'm not seeing them, can you link directly to the relevant comment(s)?

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp [hunk 8]
```diff
@@ -98,16 +98,49 @@ namespace {
   //
   // UNIT TESTS
   //
-
+#define applyAndCheckResult( A, B, newMap) \
```
[the remark this answers is not in the document]
Comment (kddevin) [reply packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp:101]: The TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL macro creates a class with members "out" and "success" that aren't accessible in a free function, so the test macros (e.g., TEST_EQUALITY) won't compile in a free function.  The #define was a cheaters work-around that I saw in other tests.  But I agree that it is distasteful.  I am replacing it with a free function that has "out" and "success" as arguments.

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp (hunk 3 above)
Comment (kddevin) [reply packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp:271]: Removed.

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp (hunk 4 above)
Comment (kddevin) [reply packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp:101]: Same solution as above.
The functions are different; one imports before the apply, the other exports after the apply.

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp (hunk 5 above)
Comment (kddevin) [reply packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp:156]: Done

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp (hunk 7 above)
Comment (kddevin) [reply packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp:271]: Done

**Comment by kddevin** [comment]
> @kddevin, the PR description mentions additional notes in re-opened issue #9391 -- somehow I'm not seeing them, can you link directly to the relevant comment(s)?

Right; it was WIP.  I added the comment to #9391 and a reproducer on branch tpetra_9391.

File: packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp [hunk 9]
```diff
@@ -4389,6 +4389,21 @@ CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
...
   }
 
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void
+  CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+  replaceDomainMap (const Teuchos::RCP<const map_type>& newDomainMap)
+  {
+    const char tfecfFuncName[] = "replaceDomainMap: ";
+    TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
+      myGraph_.is_null (), std::runtime_error,
```
[the remark this answers is not in the document]
Comment (kddevin) [reply packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp:4399]: A matrix has myGraph (nonconst) and staticGraph (const).  When a matrix is constructed from a graph, it sets only staticGraph, not myGraph, assuming the graph could be part of another matrix.  This restriction makes sense; we don't let one matrix muck with a graph that might belong to another matrix.  But it causes headaches, too, as it limits the ways we can copy a matrix and then change its domain and range maps.  

Really, the Domain and Range maps should not be in the matrix; a matrix needs only its row and column maps.  Then some other object (call it what you like, a MatrixOperator) should hold a matrix, the range and domain maps, and internal vectors laid out according to the row and column maps as needed; it should implement the apply operation.  But the four-map matrix is a design decision from Epetra that was perpetuated in Tpetra a decade ago; we'd have to think hard before redesigning it.

File: packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp [hunk 10]
```diff
@@ -4369,6 +4389,69 @@ namespace Tpetra {
...
+  {
+    const char prefix[] = "Tpetra::CrsGraph::replaceRangeMapAndExporter: ";
+    TEUCHOS_TEST_FOR_EXCEPTION(
+      rowMap_.is_null (), std::invalid_argument, prefix << "You may not call "
+      "this method unless the graph already has a column Map.");
+    TEUCHOS_TEST_FOR_EXCEPTION(
+      newRangeMap.is_null (), std::invalid_argument,
+      prefix << "The new domain Map must be nonnull.");
+
+    if (debug_) {
```
[the remark this answers is not in the document]
Comment (kddevin) [reply packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp:4427]: You are correct that this code (and, indeed, all the debug_ and verbose_ code in CrsGraph) is not tested routinely.  We need to see up tests on our machines that enable that code and all similar code.   The test could set the needed environment variables (e.g., TPETRA_DEBUG=CrsGraph).  Or we could have a mega-debug option that enables everything everywhere, and test that way.

That said, the debug statements do provide a form of documentation.  The debug checks helped me understand the assumptions made in the existing implementation of replaceDomainMapAndImporter.  So debug checks definitely have value.

And that said, the debug statements had a bug in them, which triggered failures when I enabled the environment variable.  Sigh.  The bug is fixed in the latest push.

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp (hunk 1 above)
Comment (kddevin) [reply packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp]: Yes, still relevant, but pedantic.  I removed them.

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp (hunk 2 above)
Comment (kddevin) [reply packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp]: Done

**Comment by kddevin** [comment]
Thanks for the early review, @tasmith4 .  I addressed your concerns either in code or in comments.  Let me know what you think.

File: packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (tasmith4) [reply packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp:4427]: makes sense, thanks!

File: packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (tasmith4) [reply packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp:4399]: got it, thanks

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (tasmith4) [reply packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDomainMapAndImporter.cpp]: thanks!

File: packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp (hunk 4 above)
Comment (tasmith4) [reply packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceRangeMapAndExporter.cpp]: I see, thanks

**Review by tasmith4 — APPROVED** [review]

**Pull request merged**